### ISSUE TITLE
[CI/CD] Test before merge (with github action)

### ISSUE THREAD

**Issue by ArtemisDicoTiar** [eubinecto/wisdomify#20]
젠킨스 같이 xml 혹은 yml로 어떤 조건에서 해당 테스트가 트리거 되어야하는 지 정해주면 테스트 실행후 merge가능인지 될듯

젠킨스를 이미 내 프로젝트에서 사용하고 있으니 이 프로젝트에서는 깃헙 액션으로 트리거링해보자.

**Comment by ArtemisDicoTiar** [comment]
아래 내용만 .github/workflows/ci.yml 에 추가하면 되는 거 같다.
~~~yml
name: Python application

# when triggered?
on:
  # pushed to main
  # push:
    # branches: [ main ]

  # PRd to main
  pull_request:
    branches: [ main ]

jobs:
  build:
    # OS setting
    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2 # installing python
    - name: Set up Python 3.8
      uses: actions/setup-python@v2
      with:
        python-version: 3.8
    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        pip install flake8 pytest
        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
    - name: Lint with flake8
      run: |
        # stop the build if there are Python syntax errors or undefined names
        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
        # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
    - name: Test with pytest
      run: |
        pytest
~~~

**Comment by eubinecto** [comment]
https://github.com/eubinecto/wisdomify/tree/main/wisdomify/tests

여기에 있는 테스트가 통과할 때만 머지가능하도록 저 yml 스크립트를 변경할 수는 없을까?

**Comment by ArtemisDicoTiar** [comment]
저거는 깃헙에서 제공해준 예시본이라 지금 우리 플젝 상황에 맞게 수정할게!
tests directory에 있는 pytest 파일들만 돌게끔 하면 되지?

**Comment by eubinecto** [comment]
> 저거는 깃헙에서 제공해준 예시본이라 지금 우리 플젝 상황에 맞게 수정할게!
> tests directory에 있는 pytest 파일들만 돌게끔 하면 되지?

yes! 앞으로도 쭈욱 저기에다가만 테스트를 넣을 계획

**Comment by ArtemisDicoTiar** [comment]
## action success

### main page
![Screenshot 2021-07-12 at 13 47 11](https://user-images.githubusercontent.com/47541511/125289906-aa8c3380-e317-11eb-8481-64461aa1474e.png)

### detail page
![Screenshot 2021-07-12 at 13 48 04](https://user-images.githubusercontent.com/47541511/125290033-ca235c00-e317-11eb-9127-90ff92ad4454.png)
![Screenshot 2021-07-12 at 13 51 08](https://user-images.githubusercontent.com/47541511/125290439-3736f180-e318-11eb-94c8-dd7d6af50c68.png)


pytest에서 나오던 출력 그대로 나옴. Error 가 아닌 이상 올바르게 빌드/테스트 되었다고 판단함.

## action fail
강제로 테스트 케이스 하나를 바꿔 보았다. loss.shape의 length는 0이어야하나 강제로 에러발생시키기 위해 2로 바꿈.
~~~python
    def test_training_step_dim(self):
        # (N, 3, L) -> scalar
        loss = self.rd.training_step((self.X, self.y), 0)
        self.assertEqual(len(loss.shape), 2)  # should be a scalar 
~~~

### main page
![Screenshot 2021-07-12 at 13 50 11](https://user-images.githubusercontent.com/47541511/125290315-153d6f00-e318-11eb-9e01-d92a461f2ea5.png)

### detail page
![Screenshot 2021-07-12 at 13 50 30](https://user-images.githubusercontent.com/47541511/125290365-20909a80-e318-11eb-8240-d683d9b92f31.png)

## Conclusion
잘 작동된다.
+ 참고로 깃헙 기본 템플릿에 flake8이라는 걸 추가해서 뭐가하고 찾아봤더니 여러 기능을 해주지만 여기서는 lint확인용으로 쓴다.
파이썬 파일을 올바르게 보기좋게 작성했는 지 확인 (물론 저거 올바르게 안적었다고 액션에서 에러 띄우지는 않음.)
괜찮은 기능인거 같아서 일단은 액션에 추가해놓음 빌드 중간에 "Lint with flake8"탭을 열면 어디가 파이썬 답지 않게 혹은 실행에는 문제 없으나 코드가 문제인 경우를 알려준다.
(파이썬은 인터프리터 언어라 실행하기 전까지는 에러가 어디에 어떻게 있는 지 모름. 린트는 그거를 찾아줌.)
### flake8 lint example
아래와 같은 실행에는 큰 문제가 없으나 올바르지 않게 적은 걸 알려준다.
~~~
./wisdomify/main/infer.py:36:35: F523 '...'.format(...) has unused arguments at position(s): 0
~~~
https://github.com/eubinecto/wisdomify/blob/d457e61cc47633f11add94c24ef5c2b69dd2dad8/wisdomify/main/infer.py#L36

## Finalise
- [ ] 이제 PR에서만 실행되게 수정.

**Comment by ArtemisDicoTiar** [comment]
## In addition
방금 확인한건데 에러 발생하면 이메일로도 알림이 온다. (좋다. 내가 커스텀으로 서버에 띄운 젠킨스는 그런거 하려면 플러그인 넣고 해야해서 복잡한거 같던데 ㅠㅠ)

![Screenshot 2021-07-12 at 14 01 03](https://user-images.githubusercontent.com/47541511/125292049-f5a74600-e319-11eb-9fe8-cf1a24f38734.png)

**Comment by ArtemisDicoTiar** [comment]
자 이제 PR이다. PR에서 맞게 잘 확인해주면 그대로 메인으로 머지.

**Comment by ArtemisDicoTiar** [comment]
@eubinecto pytest PR에서 실행되는 거 확인 완료.
![Screenshot 2021-07-12 at 14 05 24](https://user-images.githubusercontent.com/47541511/125292309-3e5eff00-e31a-11eb-9d57-7146ada6a0fc.png)
![Screenshot 2021-07-12 at 14 06 59](https://user-images.githubusercontent.com/47541511/125292506-6ea69d80-e31a-11eb-820e-b89121e2d688.png)

**Comment by eubinecto** [comment]
굿! 한가지 더 부탁하고 싶은것. 단순히  테스트 통과함, 아님, 이것만 체크할 수 있는게 아니라, 어떤 
테스트가 fail하는지 구체적으로 알 수 있도록 할수는 없을까>?

**Comment by ArtemisDicoTiar** [comment]
> 굿! 한가지 더 부탁하고 싶은것. 단순히 테스트 통과함, 아님, 이것만 체크할 수 있는게 아니라, 어떤
> 테스트가 fail하는지 구체적으로 알 수 있도록 할수는 없을까>?

그거 위에 액션탭가면 여태 실행된 액션 기록이 있는 데 거기에서 원하는 액션 선택하고 디테일 페이지로 가면 pytest 실행한 기록 확인 가능
즉 에러 발생하면 기존에 로컬에서 pytest실행 결과처럼 어떤 테스트가 어떤 이유로 에러 뜨는 지 확인 가능 

### 예시 (확인 방법)
![Screenshot 2021-07-12 at 14 11 44](https://user-images.githubusercontent.com/47541511/125293509-656a0080-e31b-11eb-9029-47ef75c4d920.png)
![Screenshot 2021-07-12 at 14 12 43](https://user-images.githubusercontent.com/47541511/125293507-64d16a00-e31b-11eb-87d2-c8bb024c213d.png)
![Screenshot 2021-07-12 at 14 12 57](https://user-images.githubusercontent.com/47541511/125293499-63a03d00-e31b-11eb-8210-161e71002b22.png)


#### 성공
![Screenshot 2021-07-12 at 14 10 38](https://user-images.githubusercontent.com/47541511/125293033-effe3000-e31a-11eb-9e68-b63b8e40c30a.png)

#### 실패 (에러)
![Screenshot 2021-07-12 at 14 11 15](https://user-images.githubusercontent.com/47541511/125293117-073d1d80-e31b-11eb-9881-1bf44f8d9b25.png)

**Comment by ArtemisDicoTiar** [comment]
@eubinecto 확인하고 이거로 충분하면 이슈 클로즈! 아니면 코멘트! :)

**Comment by eubinecto** [comment]
오키 충분함! 클로징 ㄱㄱ

**Comment by ArtemisDicoTiar** [comment]
### 참고
~~~
https://github.com/<OWNER>/<REPOSITORY>/workflows/<WORKFLOW_NAME>/badge.svg
~~~
흔히 보이던 배지도 이렇게 만드는 거라고 함 ㅋㅋㅋ

우리 리포의 방금 설정한 워크 플로우
https://github.com/eubinecto/wisdomify/workflows/wisdomify%20pytest%20on%20pull_request/badge.svg

나중에 리드미에 추가해도 좋을듯
